blockdevice/encryption/luks/luks.go

```diff
@@ -0,0 +1,109 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// Package luks provides a way to call LUKS2 cryptsetup.
+package luks
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"os/exec"
+	"strings"
+
+	"github.com/talos-systems/go-blockdevice/blockdevice/filesystem/luks"
+	"github.com/talos-systems/go-blockdevice/blockdevice/util"
+)
+
+// ErrEncryptionKeyRejected triggered when encryption key does not match.
+var ErrEncryptionKeyRejected error = fmt.Errorf("encryption key rejected")
+
+// LUKS implements LUKS2 encryption provider.
+type LUKS struct {
+	key    []byte
+	cipher string
+}
+
+// New creates new LUKS2 encryption provider.
+func New(key []byte, cipher string) *LUKS {
+	return &LUKS{
+		key:    key,
+		cipher: cipher,
+	}
+}
+
+// Open runs luksOpen on a device and returns mapped device path.
+func (l *LUKS) Open(deviceName string, key []byte) (string, error) {
+	parts := strings.Split(deviceName, "/")
+	mappedPath := util.PartPathEncrypted(parts[len(parts)-1])
+	parts = strings.Split(mappedPath, "/")
+	mappedName := parts[len(parts)-1]
+
+	err := runCommand([]string{"luksOpen", deviceName, mappedName}, key)
+	if err != nil {
+		return "", err
+	}
+
+	return mappedPath, nil
+}
+
+// Encrypt implements encryption.Provider.
+func (l *LUKS) Encrypt(deviceName string) error {
+	args := []string{"luksFormat", "--type", "luks2"}
+	if l.cipher != "" {
+		args = append(args, "-c", l.cipher)
+	}
+
+	args = append(args, deviceName)
+
+	err := runCommand(args, l.key)
+	if err != nil {
+		return err
+	}
+
+	return err
+}
+
+// Probe implements encryption.Provider.
+func (l *LUKS) Probe(f io.Reader) bool {
+	var header luks.SuperBlock
+
+	err := binary.Read(f, binary.LittleEndian, &header)
+	if err != nil {
+		return false
+	}
+
+	return header.Is()
+}
+
+// Close implements encryption.Provider.
+func (l *LUKS) Close(devname string) error {
+	return runCommand([]string{"luksClose", devname}, nil)
+}
+
+// runCommand executes cryptsetup with arguments.
+func runCommand(args []string, stdin []byte) error {
+	cmd := exec.Command("cryptsetup", args...)
+
+	var out bytes.Buffer
+
+	if stdin != nil {
+		cmd.Stdin = bytes.NewBuffer(stdin)
+	}
+
+	cmd.Stdout = &out
+	cmd.Stderr = &out
+
+	err := cmd.Run()
+	if err != nil {
+		if e, ok := err.(*exec.ExitError); ok && e.ExitCode() == 2 { //nolint:errorlint
+			return ErrEncryptionKeyRejected
```

Comment:
It might be useful to get the stdout and/or stderr of the command... we have a package that wraps this somewhere. I can look. Just a suggestion.

Reply:
Yeah, good point. Though in that case it won't be:
```
path, err := luks.Open(device)
if err == luks.ErrEncryptionKeyRejected {
  ...
}
```
But something like that I guess:
```
var errl luks.ErrEncrypionKeyRejected
if (errors.As(err, &errl) {
  ...
}
```

Reply:
> we have a package that wraps this somewhere. I can look.

yes, please :)